File: INE5416/prolog/fib.pl
:- dynamic fib/2.
:- retractall( fib(_,_) ).

fib(0, 0) :- !.
fib(1, 1) :- !.
fib(N, K) :-
    N1 is N-1,
    N2 is N-2,
    fib(N1, K1),
    fib(N2, K2),
    K is K1+K2,
    asserta(fib(N2, K2) :- !),
    asserta(fib(N1, K1) :- !).

File: INE5416/prolog/fib.py
# 

values = dict()
values[0] = 0
values[1] = 1

def fib(n):
    if n in values.keys():
        return values[n]
    
    f = fib(n-1) + fib(n-2)
    values[n] = f
    return f

def fib2(n):
    for i in range(n):
        fib(i)

    return fib(n)
